Update a shopcart

**As a** User
 **I need** a function to update a shopcart  
 **So that** I can update the shopcart resource in one go
   
 ### Details and Assumptions
 * We take a JSON input and update all fields at once?
   
 ### Acceptance Criteria  
   
 ```gherkin
 Given the shopcart exists
 When `PUT /shopcarts/{shopcartID}`
 Then the shopcart will be update with the provided content
 ```

I found `deserialize()` in `Item` currently not using `id` from `data`. So when I use that in my tests, something like this keeps showing up:
```
- It should update a shopcart with JSON content (FAILED)

======================================================================
FAIL: It should update a shopcart with JSON content
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/tests/test_routes.py", line 195, in test_update_shopcart
    self.assertDictEqual(resp.get_json(), shopcart.serialize())
AssertionError: {'customer_id': 68, 'id': 1, 'items': [{'co[185 chars] 1}]} != {'id': 1, 'customer_id': 68, 'items': [{'id[185 chars]w'}]}
  {'customer_id': 68,
   'id': 1,
   'items': [{'color': 'green',
-             'id': 1,
?                   ^

+             'id': 7,
?                   ^

              'name': 'mouse',
              'price': 199999.2,
              'quantity': 1,
              'shopcart_id': 1},
             {'color': 'yellow',
-             'id': 2,
?                   ^

+             'id': 8,
?                   ^

              'name': 'pc',
              'price': 199999.2,
              'quantity': 5,
              'shopcart_id': 1}]}
-------------------- >> begin captured logging << --------------------
```
So basically the only diffs are `id`s.
I add `self.id = data["id"]` to `deserialize()` and it seems working without breaking other codes. I'll include that in my PR.

--- a/service/models.py
+++ b/service/models.py
@@ -100,6 +100,7 @@ class Item(db.Model, PersistentBase):
             data (dict): A dictionary containing the resource data
         """
         try:
+            self.id = data["id"]
             self.shopcart_id = data["shopcart_id"]
             self.name = data["name"]
             self.price = data["price"]

--- a/service/routes.py
+++ b/service/routes.py
@@ -112,6 +112,38 @@ def list_all_shopcarts():
 
 
 ######################################################################
+# UPDATE A SHOPCART
+######################################################################
+@app.route("/shopcarts/<int:shopcart_id>", methods=["PUT"])
+def update_shopcart(shopcart_id):
+    """
+    Update a shopcart with JSON request.
+    Returns a 400 Error if the shopcart ID and URL does not match.
+    Returns a 200 OK with the updated content if the operation is successful.
+    """
+    app.logger.info("Updating Shopcart with id: %d", shopcart_id)
+    check_content_type("application/json")
+
+    data = request.get_json()
+    if data["id"] != shopcart_id:
+        abort(
+            status.HTTP_400_BAD_REQUEST,
+            f"Shopcart ID in JSON ({data['id']}) does not match",
+            f"the one in the request URL {shopcart_id}."
+        )
+
+    shopcart = Shopcart.find(shopcart_id)
+    if not shopcart:
+        abort(
+            status.HTTP_404_NOT_FOUND,
+            f"Shopcart with id '{shopcart_id}' could not be found."
+        )
+    shopcart.deserialize(data)
+    shopcart.update()
+    return shopcart.serialize(), status.HTTP_200_OK
+
+
+######################################################################
 # RESET A SHOPCART
 ######################################################################
 @app.route("/shopcarts/<int:shopcart_id>/reset", methods=["PUT"])

--- a/tests/test_routes.py
+++ b/tests/test_routes.py
@@ -172,3 +172,45 @@ class TestShopcartServer(TestCase):
             f"{BASE_URL}/{shopcart_id}/reset"
         )
         self.assertEqual(resp.status_code, status.HTTP_404_NOT_FOUND)
+
+    def test_update_shopcart(self):
+        """It should update a shopcart with JSON content"""
+
+        # Create a fictional shopcart and POST it
+        shopcart = self._create_shopcarts(1)[0]
+        resp = self.client.post(
+            BASE_URL, json=shopcart.serialize(), content_type="application/json"
+        )
+
+        # Create items and assign them to shopcart.items
+        items = self._create_items(randint(1, 5))
+        for item in items:
+            item.shopcart_id = shopcart.id
+        shopcart.items = items
+
+        # Update the Shopcart with JSON from shopcart
+        resp = self.client.put(
+            f"{BASE_URL}/{shopcart.id}",
+            json = shopcart.serialize(),
+            content_type="application/json"
+        )
+        # Check if successful
+        self.assertEqual(resp.status_code, status.HTTP_200_OK)
+        self.assertDictEqual(resp.get_json(), shopcart.serialize())
+
+        # Update a shopcart with shopcart_id in JSON unmatched
+        resp = self.client.put(
+            f"{BASE_URL}/{shopcart.id + randint(2,10)}",
+            json = shopcart.serialize(),
+            content_type="application/json"
+        )
+        self.assertEqual(resp.status_code, status.HTTP_400_BAD_REQUEST)
+
+        # Update a non-existing shopcart
+        shopcart.id += randint(5, 10)
+        resp = self.client.put(
+            f"{BASE_URL}/{shopcart.id}",
+            json = shopcart.serialize(),
+            content_type="application/json"
+        )
+        self.assertEqual(resp.status_code, status.HTTP_404_NOT_FOUND)

